django/db/models/fields/generated.py

```diff
@@ -0,0 +1,140 @@
+from django.core import checks
+from django.db import connections
+from django.db.models.sql import Query
+
+from . import Field
+
+__all__ = ["GeneratedField"]
+
+
+class GeneratedField(Field):
+    generated = True
+    db_returning = True
+
+    _query = None
+    _resolved_expression = None
+    output_field = None
+
+    def __init__(self, *, expression, output_field=None, db_persist=True, **kwargs):
```

Comment:
Perhaps since Oracle requires `db_persist=False` we shouldn't provide a default.

Reply:
It's a bit awkward that Postgres and Oracle are disjoint in their support. Perhaps this needs to be a three-valued thing so we can specify that we don't care if it's `STORED` or `VIRTUAL`. This might make using this in a library that supports multiple databases smoother.

Reply:
Agreed, we should do this three-valued: `None` (default), `True`, `False`.
- `True` doesn't support on Oracle,
- `False` doesn't support on PostgreSQL,
- `None` default database behavior:
  - virtual on Oracle,
  - stored on PostgreSQL,
  - virtual on SQLite (`stored` columns cannot be added with `ALTER TABLE ADD COLUMN` statement, we should add test for this),
  - virtual on MySQL.

Reply:
@jpnauta @LilyFoote What do you think? :point_up:

Reply:
Now I noticed that we don't allow any alterations to `GeneratedField`. However, there is still an issue on SQLite where `STORED` column cannot be added with `ALTER TABLE ADD COLUMN`.

Reply:
I'm OK with deferring `db_persist` to another ticket if it means this can land for Django 5.0 and we don't make it impossible to add in the future. I think it's a valuable feature to be able to choose for performance reasons, but I agree it's not something that we need to have when this lands.

Reply:
@jpnauta Do you have anything against this? If not, I can chop `db_persist` and create a new ticket when it lands (hopefully, next week).

Reply:
I just want to get a reassurance: removing `db_persist` doesn't impact the fact that this new GeneratedField will be available in PostgreSQL which implements only stored generated columns?

Reply:
If I understand correctly, we're going to use `VIRTUAL` unless it's unsupported by the backend in which case we'll use `STORED`.

Reply:
After reconsideration, I believe we need to keep `db_persist` as it will be difficult to introduce it in the future while maintaining backward compatibility. I will only change the default to `None` and force users to choose `True` or `False`.

Reply:
Done.